### PR TITLE
IN-517 separate the generate action

### PR DESCRIPTION
**Pull request by xanderflood**
# Summary

We want to start running `go generate` in Circle and checking the results in, analogous to our use of JSPrettier and Java code formatting.

To do this, we needed a separate make task that just runs `go generate` _not_ as part of other tasks, but on its own. This also required separating out `go get`, because our `go generate` tasks generally will require dependencies to have already been installed, so that they can access the tools they need.

Additionally, I'm implementing another choice made during that meeting which is to move golangci-lint into the "tools.go" pattern that we use to manage counterfeiter/mockery. The main motivation here is to move away from relying on globally-installed tools on developer machines when possible, since we lose any guarantees about having up-to-date versions in this case. This will enable our golangci-lint version to be managed by dependabot and also eliminate any discrepancies between local lints and circle lints.

I've tested this out in one Go project, but I'm not going to merge it until I've tested it out in many more. In particular, I want to make some progress on the quiz-api permissions refactor before merging.

### CODE REVIEW

File: go-common.mk
Comment (glewis-ANet) [inline go-common.mk]: Nit: I think the targets on this line were sorted?

File: go-common.mk
Comment (glewis-ANet) [inline go-common.mk]: How many other go projects should one look at?  This feels like it should be a URL which we know has information on how to do this.

File: go-common.mk
Comment (glewis-ANet) [inline go-common.mk]: We've tended to mostly name the targets after the tools they invoke (lint is perhaps an exception, although I would argue it is substantively meaningful), so should this target be 'get' instead of 'dependencies'?

This also changes behaviour a little in that test dependencies are always fetched whether we are running tests or not.  I think that is ok, but wanted see if that was of concern for anyone.

Perhaps the biggest question is that unlike the other top level targets, we're not following the same pre-/standard-/post- pattern that enables those steps to be overridden.  We've used those sparsely in our code base, but there have been cases where they've been useful.  Should we follow that pattern for these targets as well, or are we asserting that there are never useful pre or post steps for them?

**Review by glewis-ANet — COMMENTED** [review]
I think this is generally good, but had a few questions I wanted to talk through

File: go-common.mk
Comment (hcoleyAnet) [reply go-common.mk]: I have a question about the pre, standard, post thing. It seems like a bunch of those don't have anything in the pre or post targets, are those there to maintain the pattern? If so, I could see wanting to do that for this as well

File: go-common.mk
Comment (xanderflood) [reply go-common.mk]: Well by the time I finish up this ticket, every single go project should be consistent in this, but since I'm holding this for a while anyway, I'll add:

TODO: before merging, make sure this is updated to point to a specific project (this is a bit of a chicken-and-egg problem but not a big deal, I can point the link before merging the other project)

File: go-common.mk
Comment (xanderflood) [reply go-common.mk]: Hudson, I think the pre- and post-steos are all empty by design, and they're basically just "injection points" where individual projects can override to insert their own hooks.

I'll update this to follow the same pattern for "get" and "generate" - I can definitely imagine it being useful to install tools for generate (although I'd prefer we try to aim to use "get" to install our tools whenever possible, it might not always be). I can't imagine why we'd ever use it for get, but I also don't see any harm in being consistent with the other tasks.

Comment (xanderflood) [reply go-common.mk]: So re: the `-t` flag, I did a little experimenting.

First I tried removing the flag (and clearing my modcache) and found that the `go test` command actually installs the missing test dependencies on its own - we don't need to explicitly use `go get -t` beforehand.

Then I got curious and went a step further, and deleted the `get` recipe and all references to `go get` in the Makefile entirely. I went through the main recipes (test, lint, build) and repeatedly cleared my modcache before trying each one. I found is that Go will consistently just install dependencies as needed, and I wasn't able to contrive any scenario in which it was necessary to call `go get` directly. The situation doesn't appear to be any different in circle, although fully testing that out by clearing the modcache would be destructive, since I think we share that cache between builds.

@glewis-ANet do you know of any other reasons to have explicit `go get` commands, or have concerns about removing them?

Comment (glewis-ANet) [reply go-common.mk]: > So re: the `-t` flag, I did a little experimenting.
> 
> First I tried removing the flag (and clearing my modcache) and found that the `go test` command actually installs the missing test dependencies on its own - we don't need to explicitly use `go get -t` beforehand.
> 
> Then I got curious and went a step further, and deleted the `get` recipe and all references to `go get` in the Makefile entirely. I went through the main recipes (test, lint, build) and repeatedly cleared my modcache before trying each one. I found is that Go will consistently just install dependencies as needed, and I wasn't able to contrive any scenario in which it was necessary to call `go get` directly. The situation doesn't appear to be any different in circle, although fully testing that out by clearing the modcache would be destructive, since I think we share that cache between builds.
> 
> @glewis-ANet do you know of any other reasons to have explicit `go get` commands, or have concerns about removing them?

I have no objection to removing the explicit `go get` commands.  I don't see any overrides of the variables associated with that, so if we're confident that `go build` and `go test` will do the right thing without that then I'm happy with all that being gone.

File: go-common.mk
Comment (glewis-ANet) [inline go-common.mk]: A couple of concerns here.  One is that I don't see any mention of golangci in tools.go in quiz-api or in its go.mod.  Am I missing something?  The other is that this is a good idea in theory, but in practice it will mean manually intervening in all of our Go projects to pick up the new version or we'll be seeing lint failures in all of the builds.  What can we do to prevent that?  I think a transmogrifier recipe could be a solution, but wanted to know what your intentions were there?

**Review by glewis-ANet — COMMENTED** [review]
Just had a couple of questions about the golangci change

File: go-common.mk
Comment (xanderflood) [reply go-common.mk]: 1. It's been a couple months since I opened this PR, but this was addressed in [an earlier exchange we had](https://github.com/AchievementNetwork/common-make/pull/21#discussion_r1450755675). There's a chicken and egg problem since the PR (which I have open already) that adds the linter to `tools.go` is the same one that switches over to using this new common-make, so I have not merged it yet. That link will be accurate very soon, but not until after this current PR is approved and merged.

2. That's absolutely right, every go project needs to be updated. This is one of the updates I'll be making in all go projects as part of IN-517, the other main one being that they should all run `go generate` in Circle and check in the results. So more PRs will be coming, but I'm not going to open the all the other PRs until after I have approval in common-make and go-orb, in case changes need to be made here.

Comment (xanderflood) [reply go-common.mk]: (Oh I just realized that that other PR is not actually open for review yet - I'll get that open in just a minute)

Comment (glewis-ANet) [reply go-common.mk]: Sorry for the slow response here.  Thanks for the explanation.  I dug in a little more on this today and I had a concern about the use of `go mod why`, which I think from testing will also pick up indirect dependencies on golangci-lint.  I think we would like direct dependencies here.  I tried out something like this:

```
standard-lint::
	@go list -m -f '{{.Path}} {{.Indirect}}' all 2> /dev/null | \
		grep -v ' true' | \
		grep 'github.com/golangci/golangci-lint/cmd/golangci-lint' 2>&1 > /dev/null || \
		( if [ -f tools.go ]; then \
			eval "$$_common_make_tools_lint_help"; \
		  else \
			eval "$$_common_make_no_tools_lint_help"; \
		  fi; \
		  exit 1 )
	go run github.com/golangci/golangci-lint/cmd/golangci-lint run $(GOLINTFLAGS)
```

and that seemed to work.  The key there being that indirect dependencies (true) are filtered out.  The variables were defined as

```
define __common_make_no_tools_lint_help
cat <<EOF
No dependency on golangci-lint found!

Please create a top level tools.go file with contents like

-
package tools

import _ "github.com/golangci/golangci-lint/cmd/golangci-lint"
-

(the package may need adjusting) and then run 'go mod tidy'.

Then commit the new file and changes to 'go.mod' and 'go.sum'.
EOF
endef
export _common_make_no_tools_lint_help = $(value __common_make_no_tools_lint_help)

define __common_make_tools_lint_help
cat <<EOF
No dependency on golangci-lint found!

Please add an import of "github.com/golangci/golangci-lint/cmd/golangci-lint"
to the tools.go file and then run 'go mod tidy'.

Then commit the changes to the file and changes to 'go.mod' and 'go.sum'.
EOF
endef
export _common_make_tools_lint_help = $(value __common_make_tools_lint_help)
```

Comment (glewis-ANet) [reply go-common.mk]: FWIW I'm also removing the reference to a specific project there in favour of instructions about what to do

Comment (glewis-ANet) [reply go-common.mk]: I haven't tried out those instructions, but I think they're pretty close.  Please feel free to adjust them based on your experience if you decide to use that suggestion

Comment (xanderflood) [reply go-common.mk]: Thanks! The messages and bash syntax are much much clearer.

Unfortunately, that command doesn't seem to work. The command you wrote uses the `-m` flag, which only lists modules, not packages, so it will never find the package `github.com/golangci/golangci-lint/cmd/golangci-lint` no matter what the situation in our go.mod.

On the other hand, if we drop the `-m` flag and list packages, we don't actually have the `.Indirect` field to work with - that field is only defined for modules. We could possibly solve that using two commands - find the module to confirm that it's direct, and then separately search for the specific package. But even then, I'm not sure that would guarantee that the specific package is specifically direct - we could have a direct dependency, some of whose packages are only used indirectly.

Technically, both approaches are imperfect, since we want both the package, and for it to be direct. In practice though, I think either is probably fine, since I don't think it's likely that we would import part of the golangci-lint repo, or that we would import it indirectly. (For instance, stdmirror relying on go-service, which in turn lists golangci-lint as a _tools_ dependency, doesn't actually result in golangci-lint being included as an indirect dependency of stdmirror. The build constraints in the tools file help to exclude that.)

Comment (xanderflood) [reply go-common.mk]: Incidentally, I'm not able to find any way to make `go list` include dependencies mentioned in a file with build constraints. I notice that you removed the build constraints from the tools.go file, which I suspect is key to why you were able to use `go list`. But the build constraints are not optional - in general, importing an executable package in go file that isn't excluded by build constraints will cause a compiler error, especially for those projects that have non-tools go code in the same folder.

Comment (glewis-ANet) [reply go-common.mk]: Thanks Xander.  I experimented a little with this today and also couldn't see a way to either get `go list` to do what we want or to get `go mod why` to output in another format.

The one thing that I would say that might be a minor improvement is that the `go mod why` docs (https://go.dev/ref/mod#go-mod-why) don't specify the format of the error message (which we're relying on), other than it being in parentheses, but do specify a couple of other things that might be useful.  Of those, if the package *is* present as a dependency then it will display the path through the import graph to that package.  I.e., the last line of the output is going to be the package we're looking for.  Given that, we could pipe the output through `tail -1` and assert that the resulting string is equal to `github.com/golangci/golangci-lint/cmd/golangci-lint` and output the error if that isn't true.  Alternatively we could look for a line that begins and ends in parentheses as meaning it isn't present as well.

That's minor though, so I'll leave it up to you

**Review by glewis-ANet — APPROVED** [review]

File: go-common.mk
Comment (xanderflood) [reply go-common.mk]: Hmm this is tricky because you're right that the parentheses seem to be the only guarantee about that message from the docs, but they're also less meaningful and so makes the logic much more opaque IMO. I think your other suggestion - to reverse the logic and check for the path on the last line of the success output, seems like the best option. Thanks for looking into that! I checked for a structured format and then quit when I saw there wasn't one :)

**Pull request merged**
